[torch] fix exception types in custom class magic setattr/getattr

suo opened the pull request:
Summary:
`c10::AttributeError` is not automatically converted to Python AttributeError, it needs some special macros (e.g. `HANDLE_TH_ERRORS`).

Some Python functions like `hasattr` rely on the type of the throw exception to be correct.

We don't need the fully generality of those macros, so just do a targeted error type conversion here.

Test Plan: added unit test

Differential Revision: D69197217




cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

zdevito commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -785,7 +785,8 @@ void initJitScriptBindings(PyObject* module) {
                   return toPyObject(self.attr(name));
                 } catch (const ObjectAttributeError& err) {
-                  throw AttributeError("%s", err.what());
+                  pybind11::set_error(PyExc_AttributeError, err.what());
```
throw AttributeError wasn't already pybind11's way of signaling an attribute error?

suo replied:
it requires some extra boilerplate to actually translate the exception apparently, which I minimally reproduce below.

I did also try `throw pybind11::attribute_error` but for some reason this got translated into a runtime error :/

zdevito reviewed (approved):
Isn't there a pybind11 way to do this?

suo committed:
[torch] fix exception types in custom class magic setattr/getattr (#146516)

Summary:

`c10::AttributeError` is not automatically converted to Python AttributeError, it needs some special macros (e.g. `HANDLE_TH_ERRORS`).

Some Python functions like `hasattr` rely on the type of the throw exception to be correct.

We don't need the fully generality of those macros, so just do a targeted error type conversion here.

Test Plan: added unit test

Reviewed By: colin2328, zdevito

Differential Revision: D69197217
